File: assembla_cli/commands/comment.py
from __future__ import annotations

import click

from assembla_cli.formatting import print_json, print_table


@click.group()
def comment():
    """Manage ticket comments."""


@comment.command("list")
@click.argument("ticket_number", type=int)
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
@click.pass_context
def comment_list(ctx, ticket_number, as_json):
    """List comments on a ticket."""
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    data = client.get(f"/spaces/{space}/tickets/{ticket_number}/ticket_comments")
    if as_json:
        print_json(data)
    else:
        print_table(
            data or [],
            columns=["id", "author_name", "comment", "created_on"],
            headers=["ID", "Author", "Comment", "Created"],
        )


@comment.command("add")
@click.argument("ticket_number", type=int)
@click.argument("body")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
@click.pass_context
def comment_add(ctx, ticket_number, body, as_json):
    """Add a comment to a ticket.

    Usage: assembla comment add 123 "This is my comment"
    """
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    data = client.post(
        f"/spaces/{space}/tickets/{ticket_number}/ticket_comments",
        data={"ticket_comment": {"comment": body}},
    )
    if as_json:
        print_json(data)
    else:
        click.echo(f"Comment added to ticket #{ticket_number}")

File: cmd/comment.go
package cmd

import (
	"fmt"

	"github.com/eugene-software/assembla-cli/internal"
	"github.com/spf13/cobra"
)

var commentCmd = &cobra.Command{
	Use:   "comment",
	Short: "Manage ticket comments",
}

var commentListCmd = &cobra.Command{
	Use:   "list <ticket_number>",
	Short: "List comments on a ticket",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()
		asJSON, _ := cmd.Flags().GetBool("json")
		ticketNumber := args[0]

		data, err := Client.Get(internal.APIPath("spaces", space, "tickets", ticketNumber, "ticket_comments"), nil)
		if err != nil {
			return err
		}

		if asJSON {
			internal.PrintJSON(data)
		} else {
			internal.PrintTable(
				internal.ToSlice(data),
				[]string{"id", "author_name", "comment", "created_on"},
				[]string{"ID", "Author", "Comment", "Created"},
			)
		}
		return nil
	},
}

var commentAddCmd = &cobra.Command{
	Use:   "add <ticket_number> <body>",
	Short: "Add a comment to a ticket",
	Args:  cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()
		asJSON, _ := cmd.Flags().GetBool("json")
		ticketNumber := args[0]
		body := args[1]

		payload := map[string]interface{}{
			"ticket_comment": map[string]interface{}{
				"comment": body,
			},
		}

		data, err := Client.Post(internal.APIPath("spaces", space, "tickets", ticketNumber, "ticket_comments"), payload)
		if err != nil {
			return err
		}

		if asJSON {
			internal.PrintJSON(data)
		} else {
			fmt.Printf("Comment added to ticket #%s\n", ticketNumber)
		}
		return nil
	},
}

func init() {
	commentListCmd.Flags().Bool("json", false, "Output as JSON")
	commentAddCmd.Flags().Bool("json", false, "Output as JSON")

	commentCmd.AddCommand(commentListCmd)
	commentCmd.AddCommand(commentAddCmd)
}
